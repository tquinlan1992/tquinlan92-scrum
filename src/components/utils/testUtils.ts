import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { AnyAction, isType, ActionCreator } from 'typescript-fsa';
import { AppState } from '../../headless/store';

export function getAnyJestFn() {
    return jest.fn() as any;
}

interface AnyActionPayload extends AnyAction {
    payload: any;
}

export function expectActionWithPayload(actionToTest: AnyActionPayload, expectedActionCreator: ActionCreator<any>, expectedPayload?: any) {
    expect(isType(actionToTest, expectedActionCreator)).toBe(true);
    expect(actionToTest.payload).toEqual(expectedPayload);
}


const middlewares = [thunk];
export const mockStore = configureStore(middlewares);

export function getMockStore(state: Partial<AppState>) {
    return mockStore(state);
}

export function expectCalledOnceWith(mockFunction: jest.Mock<{}>, calledWith?: any) {
    expect(mockFunction).toHaveBeenCalledTimes(1);
    const expectedParams = calledWith ? [calledWith] : [];
    expect(mockFunction.mock.calls[0]).toEqual(expectedParams);
}

export function mockClearAll(mockFunctions: jest.Mock<{}>[]) {
    mockFunctions.forEach(mockFunction => {
        mockFunction.mockClear();
    })
}