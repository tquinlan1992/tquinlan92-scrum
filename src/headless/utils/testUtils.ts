export function mockPouchDB() {
    jest.mock('../database/core/pouch', () => {
        return {};
    });
}
