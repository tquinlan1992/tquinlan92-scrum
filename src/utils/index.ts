import * as _ from 'lodash';
import { connect, MapStateToPropsParam, MergeProps } from 'react-redux';
import { AppState } from '@headless/store';
import { ThunkDispatch } from 'redux-thunk';
import { AnyAction } from 'redux';

type PropsMap<M, K extends keyof M> = {[P in K]: M[P]};

export function pick<M extends object, K extends keyof M>(obj: M, ...props: K[]): PropsMap<M, K> {
    // @ts-ignore: TS2352: Type 'PartialDeep<M>' cannot be converted to type 'PropsMap<M, K>'.
    return _.pick<M>(obj, props) as PropsMap<M, K>;
}

export type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];

export type AppStateThunkDispatch = ThunkDispatch<AppState, undefined, AnyAction>;
