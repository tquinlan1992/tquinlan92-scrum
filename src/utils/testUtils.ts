export function mockPouchDB() {
    jest.mock('../headless/database/pouch', () => {
        return {};
    });
}

export function mockPouch(dbMethods: object) {
    jest.mock('../headless/database/pouch', () => {
        return {

        }
    })
    const mockPouchRequire = require('../headless/database/pouch');
    const mockGetRemoteDb = jest.fn(() => {
        return new Promise(resolve => {
            resolve(dbMethods);
        })
    });
    mockPouchRequire.getRemoteDb = mockGetRemoteDb;
    return mockPouchRequire;
}

export function mockShortId() {
    jest.mock('shortid', () => {
        let count = 0;
        return () => {
            count = count + 1;
            return String(count);
        }
    })
}

