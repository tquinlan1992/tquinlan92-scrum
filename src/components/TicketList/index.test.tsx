import * as React from 'react';
import { shallow, configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import { TicketList } from './';
import { forEach, get } from 'lodash';
import { getAnyJestFn } from '../../utils/testUtils';

configure({ adapter: new Adapter() });

const mockActions = {
    setShowAddTicketDialog: getAnyJestFn(),
    fetchTickets: getAnyJestFn(),
    addTicket: getAnyJestFn(),
    closeTicket: getAnyJestFn(),
    openAddTicketDialog: getAnyJestFn(),
    setTicketListState: getAnyJestFn(),
    onRemoveFromSprint: getAnyJestFn(),
    closeSprint: getAnyJestFn(),
    openCloseSprintDialog: getAnyJestFn(),
    updatePriorities: getAnyJestFn()
};

function testShowAddTicketDialogValue(showAddTicketDialog: boolean) {
    describe(`showAddTicketDialog as ${showAddTicketDialog}`, () => {
        it(`TicketList should show the addTicketDialog open as ${showAddTicketDialog}`, () => {
            const props = {
                showAddTicketDialog,
                showCloseSprintDialog: false,
                setShowAddTicketDialog: mockActions.setShowAddTicketDialog as any,
                fetchTickets: mockActions.fetchTickets as any,
                addTicket: mockActions.addTicket as any,
                closeTicket: mockActions.addTicket as any,
                addTicketToSprint: mockActions.addTicket as any,
                setTicketListState: mockActions.setTicketListState as any,
                setAddTicketDialogState: mockActions.setTicketListState as any,
                openAddTicketDialog: mockActions.openAddTicketDialog,
                onRemoveFromSprint: mockActions.onRemoveFromSprint,
                closeSprint: mockActions.closeSprint,
                openCloseSprintDialog: mockActions.openCloseSprintDialog,
                updatePriorities: mockActions.updatePriorities,
                backlogTickets: [{ title: 'backlogTicketTitle', description: 'backlogDescription', storyPoint: 3, _id: 'id1', deleted: false} as any],
                sprintTickets: [{ title: 'sprintTicketTitle', description: 'sprintDescription', storyPoint: 3, _id: 'id1', deleted: false} as any],
                closedTickets: [{ title: 'closedTicketTitle', description: 'closedDescription', storyPoint: 3, _id: 'id1', deleted: false} as any]
            };
            const result = shallow(<TicketList {...props} />);
            expect(result).toMatchSnapshot();
        });
    });
}

describe('describe when the TicketList Component is given', () => {
        testShowAddTicketDialogValue(true);
        testShowAddTicketDialogValue(false);
});

describe('when a user', () => {
    beforeEach(() => {
        forEach(mockActions, mockAction => {
            mockAction.mockClear();
        });
    });
    describe('requests to close addTicket', () => {
        it('setShowAddTicketDialog should be called with false', () => {
            const props = {
                showAddTicketDialog: false,
                showCloseSprintDialog: false,
                setShowAddTicketDialog: mockActions.setShowAddTicketDialog as any,
                fetchTickets: mockActions.fetchTickets as any,
                addTicket: mockActions.addTicket as any,
                closeTicket: mockActions.addTicket as any,
                addTicketToSprint: mockActions.addTicket as any,
                setTicketListState: mockActions.setTicketListState as any,
                setAddTicketDialogState: mockActions.setTicketListState as any,
                openAddTicketDialog: mockActions.openAddTicketDialog,
                onRemoveFromSprint: mockActions.onRemoveFromSprint,
                closeSprint: mockActions.closeSprint,
                openCloseSprintDialog: mockActions.openCloseSprintDialog,
                updatePriorities: mockActions.updatePriorities,
                backlogTickets: [{ title: 'backlogTicketTitle', description: 'backlogDescription', storyPoint: 3, _id: 'id1', deleted: false} as any],
                sprintTickets: [{ title: 'sprintTicketTitle', description: 'sprintDescription', storyPoint: 3, _id: 'id1', deleted: false} as any],
                closedTickets: [{ title: 'closedTicketTitle', description: 'closedDescription', storyPoint: 3, _id: 'id1', deleted: false} as any]
            };
            const result = shallow(<TicketList {...props} />);
            const dialogProps: any = result.find('Connect(AddTicketDialog)').props();
            dialogProps.onRequestClose();
            expect(props.setAddTicketDialogState.mock.calls).toMatchObject([
                [{ open: false }]
            ]);
        });
    });
});
