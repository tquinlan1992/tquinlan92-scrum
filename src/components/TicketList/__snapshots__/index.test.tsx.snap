// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`describe when the TicketList Component is given showAddTicketDialog as false TicketList should show the addTicketDialog open as false 1`] = `
<div>
  <WithStyles(Paper)
    style={
      Object {
        "margin": "5px auto",
      }
    }
  >
    <TicketTable
      tickets={
        Array [
          Object {
            "_id": "id1",
            "deleted": false,
            "description": "closedDescription",
            "storyPoint": 3,
            "title": "closedTicketTitle",
          },
        ]
      }
      title="Closed"
    />
  </WithStyles(Paper)>
  <WithStyles(Paper)
    style={
      Object {
        "margin": "5px auto",
      }
    }
  >
    <TicketTable
      onClose={[Function]}
      onRemoveFromSprint={[Function]}
      tickets={
        Array [
          Object {
            "_id": "id1",
            "deleted": false,
            "description": "sprintDescription",
            "storyPoint": 3,
            "title": "sprintTicketTitle",
          },
        ]
      }
      title="Sprint"
    />
    <WithStyles(Button)
      onClick={[Function]}
      title="Close Sprint"
    >
       Close Sprint 
    </WithStyles(Button)>
  </WithStyles(Paper)>
  <WithStyles(Paper)
    style={
      Object {
        "margin": "5px auto",
      }
    }
  >
    <DraggableList
      listItems={
        Array [
          Object {
            "_id": "id1",
            "deleted": false,
            "description": "backlogDescription",
            "storyPoint": 3,
            "title": "backlogTicketTitle",
          },
        ]
      }
      title="Backlog"
      updateItems={[Function]}
    />
    <WithStyles(Button)
      onClick={[Function]}
      title="Add Ticket"
    >
       Add Ticket 
    </WithStyles(Button)>
  </WithStyles(Paper)>
  <Connect(AddTicketDialog)
    onRequestClose={[Function]}
    onSubmit={[Function]}
    open={false}
  />
  <Connect(CloseSprintDialog)
    onRequestClose={[Function]}
    onSubmit={[Function]}
    open={false}
  />
</div>
`;

exports[`describe when the TicketList Component is given showAddTicketDialog as true TicketList should show the addTicketDialog open as true 1`] = `
<div>
  <WithStyles(Paper)
    style={
      Object {
        "margin": "5px auto",
      }
    }
  >
    <TicketTable
      tickets={
        Array [
          Object {
            "_id": "id1",
            "deleted": false,
            "description": "closedDescription",
            "storyPoint": 3,
            "title": "closedTicketTitle",
          },
        ]
      }
      title="Closed"
    />
  </WithStyles(Paper)>
  <WithStyles(Paper)
    style={
      Object {
        "margin": "5px auto",
      }
    }
  >
    <TicketTable
      onClose={[Function]}
      onRemoveFromSprint={[Function]}
      tickets={
        Array [
          Object {
            "_id": "id1",
            "deleted": false,
            "description": "sprintDescription",
            "storyPoint": 3,
            "title": "sprintTicketTitle",
          },
        ]
      }
      title="Sprint"
    />
    <WithStyles(Button)
      onClick={[Function]}
      title="Close Sprint"
    >
       Close Sprint 
    </WithStyles(Button)>
  </WithStyles(Paper)>
  <WithStyles(Paper)
    style={
      Object {
        "margin": "5px auto",
      }
    }
  >
    <DraggableList
      listItems={
        Array [
          Object {
            "_id": "id1",
            "deleted": false,
            "description": "backlogDescription",
            "storyPoint": 3,
            "title": "backlogTicketTitle",
          },
        ]
      }
      title="Backlog"
      updateItems={[Function]}
    />
    <WithStyles(Button)
      onClick={[Function]}
      title="Add Ticket"
    >
       Add Ticket 
    </WithStyles(Button)>
  </WithStyles(Paper)>
  <Connect(AddTicketDialog)
    onRequestClose={[Function]}
    onSubmit={[Function]}
    open={true}
  />
  <Connect(CloseSprintDialog)
    onRequestClose={[Function]}
    onSubmit={[Function]}
    open={false}
  />
</div>
`;
